### PR TITLE
(#109) fix: wildcard route

### PR DESCRIPTION
**Pull request by dannypsnl**

### CODE REVIEW

**Comment by dannypsnl** [comment]
@shana0440 You can review it now.

**Comment by dannypsnl** [comment]
will, I think I just fix with refactor, don't merge it

**Comment by dannypsnl** [comment]
And #105

File: filler.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	return f.nextFill(ctx)
+}
+
+func newRouteFiller(h *handler, rs []string, next contextFiller) contextFiller {
+	r := &routeFiller{h: h, rs: rs}
+	r.next = next
+	return r
+}
+func newQueryFiller(h *handler, query url.Values, next contextFiller) contextFiller {
+	r := &queryFiller{h: h, query: query}
```
Comment (dannypsnl) [inline filler.go]: `r` is inconsistent naming, should be `q`, also other constructor

File: filler.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+			value := parseParameter(field, p)
+			field.Set(value)
+		}
+	}
+	return q.nextFill(ctx)
+}
+func (j *jsonFiller) fill(ctx reflect.Value) error {
+	if j.h.expectJsonRequest {
+		v := ctx.Interface()
+		err := json.NewDecoder(j.req.Body).Decode(v)
```
Comment (dannypsnl) [inline filler.go]: Try to send `req.Body` here, not `req`

File: filler.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		if v, ok := f.form[k]; ok {
+			field := ctx.Elem().Field(idx)
+			p := v[0]
+			value := parseParameter(field, p)
+			field.Set(value)
+		}
+	}
+	return f.nextFill(ctx)
+}
+
```
Comment (dannypsnl) [inline filler.go]: Better to add default filler chain, users might break the assuming, route -> query -> json -> form is required!

File: handler.go [hunk 4]
```diff
@@ -82,53 +82,18 @@ func (h *handler) context(rs []string, req *http.Request) []reflect.Value {
...
+		filler := newRouteFiller(h, rs,
+			newQueryFiller(h, req.URL.Query(),
+				newJSONFiller(h, req,
+					newFormFiller(h, req.Form, nil),
+				),
+			),
+		)
+		err := filler.fill(context)
+		if err != nil {
+			param[h.userDefinedContextOffset] = reflect.ValueOf(errors.New("400"))
```
Comment (dannypsnl) [inline handler.go]: bug

Comment (dannypsnl) [reply handler.go]: Might not fix it here, I think I fix this at #86

File: router.go [hunk 5]
```diff
@@ -41,13 +41,14 @@ func (route *Route) addHandlerTo(routeStr string, h *handler) {
...
 	child := route.Children
-	for _, r := range rs {
+	for i, r := range rs {
 		if isParameter(r) {
 			if matchRoute.VariableRoute == nil {
 				matchRoute.VariableRoute = NewRoute()
 			}
 			matchRoute = matchRoute.VariableRoute
 		} else if r[0] == '*' {
+			h.matchedPathIndex = i - (len(routes) - len(h.routes))
```
Comment (dannypsnl) [inline router.go]: confusing

File: server_test.go [hunk 6]
```diff
@@ -32,6 +32,11 @@ type (
 	Files struct {
```
Comment (dannypsnl) [inline server_test.go]: I think is time to split this huge test... it's too big!

File: router.go (hunk 5 above)
Comment (dannypsnl) [reply router.go]: Not sure how to fix

File: server_test.go (hunk 6 above)
Comment (dannypsnl) [reply server_test.go]: I think don't block this PR by this, fix at other branch

File: router.go (hunk 5 above)
Comment (dannypsnl) [reply router.go]: @shana0440 Any ideas?

Comment (dannypsnl) [reply router.go]: `len(routes) - len(h.routes)` means `baseRouteLen`

File: filler.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+func defaultFillerChain(h *handler, rs []string, body io.Reader, query, form url.Values) contextFiller {
+	return newRouteFiller(h, rs,
+		newQueryFiller(h, query,
+			newJSONFiller(h, body,
+				newFormFiller(h, form, nil),
+			),
+		),
+	)
+}
```
Comment (shana0440) [inline filler.go]: I prefer this way
```
type Chain struct {
  context reflect.Value
  error errors
}

func (c *Chain) pipe (f *filler) {
  if c.error == nil {
    c.error = f.fill(c.context);
  }
}

err := newChain(context).
  pipe(newRouteFiller(h, routes)).
  pipe(newQueryFiller(h, query)).
  pipe(newJSONFiller(h, body)).
  pipe(newFormFiller(h, form)).
  error
```

File: router.go (hunk 5 above)
Comment (dannypsnl) [reply router.go]: Yep then I think we should remove matchedPathIndex at other place

File: filler.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	return nil
+}
+func (r *routeFiller) fill(ctx reflect.Value) error {
+	for idx, route := range r.h.routes {
+		if isParameter(route) {
+			param := r.rs[len(r.rs)-len(r.h.routes)+idx]
+			index := r.h.routeParams[idx]
+			v := parseParameter(ctx.Elem().Field(index), param)
+			ctx.Elem().Field(index).
+				Set(v)
```
[the remark this answers is not in the document]
Comment (dannypsnl) [reply filler.go]: My purpose is making it more obvious.

File: filler.go (hunk 7 above)
Comment (dannypsnl) [reply filler.go]: Has a problem, JSON filler would stop form filler if it matched

Comment (dannypsnl) [reply filler.go]: Or change fill interface, but I think both are over design now.

Comment (shana0440) [reply filler.go]: but right now, `defaultFillerChain` need to know every argument of every filler, if we want to create a new filler like `binaryFiller` for example, then we need to change `defaultFillerChain` and its caller, our arguments hell grow also, so I think is better to fix it now.

Comment (dannypsnl) [reply filler.go]: I move default chain into handler, would it make sense about let it know what info should handler know now?

**Comment by dannypsnl** [comment]
@shana0440 If looks good to you, can you merge this?

**Pull request merged**